### PR TITLE
wip: Multi input output for DL

### PR DESCRIPTION
**Pull request by orhankislal**
- [ ] Add the module name, JIRA# to PR/commit and description.
- [ ] Add tests for the change.

### CODE REVIEW

File: src/ports/postgres/modules/deep_learning/input_data_preprocessor.py_in [hunk 1]
```diff
@@ -89,12 +89,15 @@ class InputDataPreprocessorDL(object):
...
+            for i in range(len(self.dependent_levels)):
+                self.dependent_levels[i] = ['NULL' if level is None else level
+                    for level in self.dependent_levels[i]]
             self._validate_num_classes()
             # Try computing padding_size after running all necessary validations.
+            self.padding_size = []
             if self.num_classes:
-                self.padding_size = self.num_classes - len(self.dependent_levels)
+                for i in range(len(self.dependent_levels)):
+                    self.padding_size[i] = self.num_classes - len(self.dependent_levels[i])
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/input_data_preprocessor.py_in]: I'm a little confused about this.  Why is there only 1 num_classes, but many different dependent_levels lists and many different padding sizes?  If i here is iterating over different outputs, shouldn't there be a different num_classes for each i?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 2]
```diff
@@ -109,13 +101,32 @@ def fit(schema_madlib, source_table, model, model_arch_table,
...
+    columns_dict['mb_indep_var_cols'] = src_summary_dict['independent_varname']
+
+    columns_dict['dep_shape_cols'] = [add_postfix(i, "_shape") for i in columns_dict['mb_dep_var_cols']]
+    columns_dict['ind_shape_cols'] = [add_postfix(i, "_shape") for i in columns_dict['mb_indep_var_cols']]
+
+    fit_validator.dependent_varname = columns_dict['mb_dep_var_cols']
+    fit_validator.independent_varname = columns_dict['mb_indep_var_cols']
+    fit_validator.dep_shape_col = columns_dict['dep_shape_cols']
+    fit_validator.ind_shape_col = columns_dict['ind_shape_cols']
+
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:147]: We may as well rename dep_shape_col in FitInputValidator to dep_shape_cols as well--keep things consistent.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 3]
```diff
@@ -311,7 +332,7 @@ def fit(schema_madlib, source_table, model, model_arch_table,
             $5 AS metrics_elapsed_time,
             '{version}'::TEXT AS madlib_version,
             {num_classes}::INTEGER AS num_classes,
-            $6 AS {class_values_colname},
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:314]: Since class_values_colnames was removed here, we should also remove it from the list of params passed to format.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 4]
```diff
@@ -246,14 +265,16 @@ def fit(schema_madlib, source_table, model, model_arch_table,
     end_training_time = datetime.datetime.now()
 
     version = madlib_version(schema_madlib)
-    src_summary_dict = get_source_summary_table_dict(fit_validator)
-    class_values = src_summary_dict['class_values']
+    class_values = []
+    for i in class_values_colnames:
+        class_values.append("$MAD${0}$MAD$ AS {1}".format(src_summary_dict[i], i))
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: We should probably start getting in the habit of using quote_literal() and friends more.  (Or quote_nullable, if the class names are allowed to be NULL?)
   "{0} AS {1}".format(quote_literal(src_summary_dict[i]), i))

might make debugging the SQL query easier if we have to print it out later.  Instead of having a string like '$MAD$value1$MAD$ AS colname,$MAD$value2$MAD$ AS colname, $MAD$value3$MAD$...'
it would just be 'value1 AS colname','value2' AS colname,'value3' AS colname with any ' characters inside the strings replaces with ''

At least for gpdb6, it's built in to plpy, not sure about gpdb5?

Also... maybe quote_ident(i) instead of i?  Not sure if that's already been quoted somewhere earlier.

File: src/ports/postgres/modules/deep_learning/madlib_keras.sql_in [hunk 5]
```diff
@@ -1648,13 +1648,13 @@ CREATE OR REPLACE FUNCTION MADLIB_SCHEMA.madlib_keras_fit(
...
     fit_params              VARCHAR,
     num_iterations          INTEGER,
-    use_gpus                BOOLEAN,
-    validation_table        VARCHAR,
-    metrics_compute_frequency  INTEGER,
-    warm_start              BOOLEAN,
-    name                    VARCHAR,
-    description             VARCHAR,
-    object_table            VARCHAR
+    use_gpus                BOOLEAN DEFAULT FALSE,
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.sql_in:1651]: Will this work in gpdb5?

File: src/ports/postgres/modules/deep_learning/madlib_keras_validator.py_in [hunk 6]
```diff
@@ -269,8 +272,8 @@ class FitCommonValidator(object):
         self.model_id = model_id
         self.dependent_varname = dependent_varname
         self.independent_varname = independent_varname
-        self.dep_shape_col = add_postfix(dependent_varname, "_shape")
-        self.ind_shape_col = add_postfix(independent_varname, "_shape")
+        # self.dep_shape_col = add_postfix(dependent_varname, "_shape")
+        # self.ind_shape_col = add_postfix(independent_varname, "_shape")
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras_validator.py_in]: Remove?

File: src/ports/postgres/modules/deep_learning/madlib_keras_validator.py_in [hunk 7]
```diff
@@ -284,7 +287,7 @@ class FitCommonValidator(object):
                 self.output_model_table, "_summary")
         self.accessible_gpus_for_seg = accessible_gpus_for_seg
         self.module_name = module_name
-        self._validate_common_args()
+        # self._validate_common_args()
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras_validator.py_in]: If we're not going to call this, should we remove the method from this class?

File: src/ports/postgres/modules/deep_learning/model_arch_info.py_in [hunk 8]
```diff
@@ -37,8 +37,12 @@ def _get_layers(model_arch):
...
     arch_layers = _get_layers(model_arch)
-    if 'batch_input_shape' in arch_layers[0]['config']:
-        return arch_layers[0]['config']['batch_input_shape'][1:]
+    shapes = []
+    for i in arch_layers:
+        if 'batch_input_shape' in i['config']:
+            shapes.append(i['config']['batch_input_shape'][1:])
+    if shapes is not None:
+        return shapes
     plpy.error('Unable to get input shape from model architecture.')
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/model_arch_info.py_in:46]: Should be:
```
  if not shapes:
       return shapes
```

File: src/ports/postgres/modules/utilities/utilities.py_in [hunk 9]
```diff
@@ -1106,19 +1106,23 @@ def validate_module_input_params(source_table, output_table, independent_varname
...
+
+
+
+    for i in dependent_varname:
+        _assert(is_var_valid(source_table, i),
+                "{module_name} error: invalid dependent_varname "
+                "('{dependent_varname}') for source_table "
+                "({source_table})!".format(module_name=module_name,
+                                           dependent_varname=i,
+                                           source_table=source_table))
```
Comment (reductionista) [inline src/ports/postgres/modules/utilities/utilities.py_in:1128]: Won't this change break other MADlib modules that call it (linear regression, mlp, etc...), passing strings for independent_varname and dependent_varname?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 10]
```diff
@@ -386,26 +407,13 @@ def get_initial_weights(model_table, model_arch, serialized_weights, warm_start,
...
-                   indep_varname='independent_varname',
-                   tbl=fit_validator.source_summary_table))[0]
-    source_summary['class_values_type'] = get_expr_type(
-        CLASS_VALUES_COLNAME, fit_validator.source_summary_table)
-    source_summary['norm_const_type'] = get_expr_type(
-        NORMALIZING_CONST_COLNAME, fit_validator.source_summary_table)
+            SELECT *
+            FROM {0}
+        """.format(fit_validator.source_summary_table))[0]
+
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:463]: Maybe the param for this `get_source_summary_table_dict()` helper should just be `source_summary_table` instead of `fit_validator` since that's the only thing we use out of it.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 11]
```diff
@@ -164,12 +176,19 @@ def fit(schema_madlib, source_table, model, model_arch_table,
         # with the function definition
         plpy.error("Object table not specified for function {0} in compile_params".format(custom_fn_list))
 
+    mb_dep_var_cols_sql = ', '.join(["dependent_var_{0}".format(i)
+                                    for i in columns_dict['mb_dep_var_cols']])
+    mb_indep_var_cols_sql = ', '.join(["independent_var_{0}".format(i)
+                                    for i in columns_dict['mb_indep_var_cols']])
+    dep_shape_cols_sql = ', '.join(columns_dict['dep_shape_cols'])
+    ind_shape_cols_sql = ', '.join(columns_dict['ind_shape_cols'])
+
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:235]: quote_ident(dependent_var_{0}.format(i)) etc.  might be good here too... I think it would only matter if we want to allow " or \ in the column names though.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 12]
```diff
@@ -511,17 +516,23 @@ def fit_transition(state, dependent_var, independent_var, dependent_var_shape,
...
+        y_train.append(np_array_int16(dependent_var[counter], shape))
 
     # Fit segment model on data
     #TODO consider not doing this every time
     fit_params = parse_and_validate_fit_params(fit_params)
     segment_model.fit(x_train, y_train, **fit_params)
 
     # Aggregating number of images, loss and accuracy
-    agg_image_count += len(x_train)
+    agg_image_count += len(x_train[0])
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:613]: Maybe we should add an assert here, that len(x[0]) == len(x_train[0]) for x in x_train[1:]

Or I guess better, just check it once in the validator (if it's not there already).

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 13]
```diff
@@ -903,9 +948,15 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
...
-    state[1] = agg_metric
-    state[2] = agg_image_count
+    state = [agg_loss]
+
+    if multi_output:
+        for counter in range(output_count):
+            state.append(aux_losses[counter])
+            state.append(aux_metrics[counter])
+    else:
+        state[1] = agg_metric
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: I think this will give an out of range exception.  
  Use  `state.append(agg_metric)`

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 14]
```diff
@@ -849,7 +859,25 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
                                    is_final_iteration, custom_function_map=None, **kwargs):
     SD = kwargs['SD']
     device_name = get_device_name_and_set_cuda_env(accessible_gpus_for_seg[current_seg_id], current_seg_id)
-    agg_loss, agg_metric, agg_image_count = state
+
+    multi_output = True if type(dependent_var) is list and len(dependent_var) > 1 else False
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:965]: I think it would be better if we just check for `type(dependent_var) is list` (without the second condition), and then at the start of the `if multi_output:` branch, assert(len(dependent_var) > 1)

Seems like if `dependent_var` is a 1-element list at this point, and we don't treat it like multi-output, it's probably going to hit an exception anyway if it gets treated like a string instead of a list.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 15]
```diff
@@ -873,25 +901,42 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
...
-    x_val = np_array_float32(independent_var, independent_var_shape)
-    y_val = np_array_int16(dependent_var, dependent_var_shape)
+    x_val = []
+    y_val = []
+    for counter, shape in enumerate(independent_var_shape):
+
+        x_val.append(np_array_float32(independent_var[counter], shape))
+    for counter, shape in enumerate(dependent_var_shape):
+
+        y_val.append(np_array_int16(dependent_var[counter], shape))
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:1011]: Need another branch here for `multi_output = False` case

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 16]
```diff
@@ -873,25 +901,42 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
...
-        loss, metric = res
+    if multi_output:
+        loss = res[0]
+        agg_loss += (image_count * loss)
+        for counter in range(output_count):
+            # For multi output cases, res has the following structure
+            # print(model.metrics_names)
+            # ['loss', 'dense_4_loss', 'dense_5_loss', 'dense_4_acc', 'dense_5_acc']
+            aux_losses[counter] = aux_losses[counter] + (image_count * res[counter+1])
+            aux_metrics[counter] = aux_metrics[counter] + (image_count * res[counter+1+len(dependent_var)])
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:1029]: Wow, didn't realize eval was going to get this complicated!  Looks right though.

**Review by reductionista — COMMENTED** [review]
Coming along well!

Needs some more validation code filled in, and better handling of the single-output case... which I think you're already working on.

My main concern is with num_classes.  It seems like we might want to change the interface, so that the user can pass a max number of classes for each output.  

Let's say the model has one main output that can take 365 values, and a second output that's just a boolean flag, only 2 values.  If I'm reading the current code right, then it seems like the minibatcher would take num_classes=365 to mean that we should use a 365-element array just to hold the boolean output, in other words... padding it with 363 zeros.  Plus the other one which would be another 365-element array.  So what should have been able to fit in 367 elements instead takes up twice as much space.

I'm not sure what the best interface is, but seems like the user ought to be able to pass a list like 2,365 to num_classes.

File: src/ports/postgres/modules/deep_learning/model_arch_info.py_in (hunk 8 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/model_arch_info.py_in:46]: This logic is the opposite of what we want. We return shapes if it does exist.

Comment (reductionista) [reply src/ports/postgres/modules/deep_learning/model_arch_info.py_in:46]: Yes, sorry... I meant to write:
```
   if shapes:
        return shapes
```
The current logic will always return True, since it cannot be None

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 12 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in:613]: Checking the counts for every single row should not be necessary. If the counts did not match for some reason, the `fit` would fail before this point with a proper keras error.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 14 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in:965]: 1 element list is acceptable as a single output.

File: src/ports/postgres/modules/deep_learning/madlib_keras.sql_in (hunk 5 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras.sql_in:1651]: It does work with gpdb5

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 14 above)
Comment (reductionista) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in:965]: Oh!  I see, so dependent_var is always a list.
Maybe we don't need to check whether it's a list then?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 15 above)
Comment (reductionista) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in:1011]: Nevermind about this, had the same confusion here.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 17]
```diff
@@ -298,13 +324,13 @@ def fit(schema_madlib, source_table, model, model_arch_table,
...
         # in the else part.
-        validation_table = "$MAD${0}$MAD$".format(validation_table)
+        validation_table = quote_literal(validation_table)
     else:
         validation_metrics = validation_loss = 'NULL'
         validation_metrics_final = validation_loss_final = 'NULL'
         validation_table = 'NULL'
 
-    object_table = "$MAD${0}$MAD$".format(object_table) if object_table is not None else 'NULL'
+    object_table = quote_literal(object_table) if object_table is not None else 'NULL'
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:356]: I think this should be quote_ident(object_table)

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 18]
```diff
@@ -736,7 +789,7 @@ def evaluate(schema_madlib, model_table, test_table, output_table,
     compile_params_query = "SELECT compile_params, metrics_type, object_table FROM {0}".format(model_summary_table)
     res = plpy.execute(compile_params_query)[0]
     metrics_type = res['metrics_type']
-    compile_params = "$madlib$" + res['compile_params'] + "$madlib$"
+    compile_params = quote_literal(res['compile_params'])
     object_table = res['object_table']
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in:831]: quote_ident(object_table)

File: src/ports/postgres/modules/deep_learning/test/madlib_keras_iris.setup.sql_in [hunk 19]
```diff
@@ -293,3 +293,36 @@ SELECT train_test_split('iris_data',     -- Source table
...
+$$::json,  NULL,
+ 'Sophie',
+                               'A simple model'
```
Comment (reductionista) [inline src/ports/postgres/modules/deep_learning/test/madlib_keras_iris.setup.sql_in:310]: This reminds me of something I'd been thinking for a while--it seems like in these dev-check tests we should be starting from python code rather than json, or at least check in the python code somewhere that generated the json.

Otherwise, every time the keras or tensorflow version increases, we'll have to re-generate the json from python code, and if it's not checked in then we may not even be able to find the code, and have to start from scratch again.

Probably a separate story though.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 18 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in:831]: We need literal text for these variables. quote_ident gives us " enclosing which looks like a column for postgresql commands.
